l10n_eu_product_adr_delivery_carrier/models/stock_picking.py

```diff
@@ -0,0 +1,28 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import api, models
+
+
+class StockPicking(models.Model):
+
+    _inherit = "stock.picking"
+
+    @api.onchange("carrier_id")
+    def _onchange_carrier_id(self):
+        self._check_adr()
+
+    def _check_adr(self):
+        """Display a warning if any move has a product with ADR settings
+        matching those defined on the carrier"""
+        warnings = {}
+        if self.carrier_id and self.carrier_id.adr_limited_amount_ids:
+            for move in self.move_lines:
+                if (
+                    move.product_id.limited_amount_id
+                    in self.carrier_id.adr_limited_amount_ids
+                ):
+                    warnings[move] = move.product_id.limited_amount_id
+                # Other "ADR" M2m fields to be defined on carrier
```

Comment:
you could add `_adr_validators` (or similar) which returns the callables for these checks. 
Then it's easy to add more validators w/out touching this code.